pkg/server/list_users.go

```diff
@@ -0,0 +1,131 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/openfga/openfga/internal/utils"
+
+	grpc_ctxtags "github.com/grpc-ecosystem/go-grpc-middleware/tags"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/openfga/openfga/internal/condition"
+	"github.com/openfga/openfga/internal/graph"
+	"github.com/openfga/openfga/pkg/middleware/validator"
+	"github.com/openfga/openfga/pkg/telemetry"
+
+	"github.com/openfga/openfga/pkg/server/commands/listusers"
+	serverErrors "github.com/openfga/openfga/pkg/server/errors"
+	"github.com/openfga/openfga/pkg/typesystem"
+)
+
+// ListUsers returns all subjects (users) of a specified terminal type
+// that are relate via specific relation to a specific object.
+func (s *Server) ListUsers(
+	ctx context.Context,
+	req *openfgav1.ListUsersRequest,
+) (*openfgav1.ListUsersResponse, error) {
+	start := time.Now()
+	ctx, span := tracer.Start(ctx, "ListUsers", trace.WithAttributes(
+		attribute.String("store_id", req.GetStoreId()),
+		attribute.String("object", fmt.Sprintf("%s:%s", req.GetObject().GetType(), req.GetObject().GetId())),
+		attribute.String("relation", req.GetRelation()),
+		attribute.String("user_filters", userFiltersToString(req.GetUserFilters())),
+	))
+	defer span.End()
+	if !s.IsExperimentallyEnabled(ExperimentalEnableListUsers) {
+		return nil, status.Error(codes.Unimplemented, "ListUsers is not enabled. It can be enabled for experimental use by passing the `--experimentals enable-list-users` configuration option when running OpenFGA server")
+	}
+
+	if !validator.RequestIsValidatedFromContext(ctx) {
+		if err := req.Validate(); err != nil {
+			return nil, status.Error(codes.InvalidArgument, err.Error())
+		}
+	}
+
+	const methodName = "listusers"
+
+	typesys, err := s.resolveTypesystem(ctx, req.GetStoreId(), req.GetAuthorizationModelId())
+	if err != nil {
+		telemetry.TraceError(span, err)
+		return nil, err
+	}
+
+	err = listusers.ValidateListUsersRequest(ctx, req, typesys)
+	if err != nil {
+		telemetry.TraceError(span, err)
```

Comment:
Is an invalid request an error per se? It's not really. Nothing went wrong. The user just gave us a bad request. This will turn a lot of spans red in our Trace Dashboard even if there wasn't anything really wrong with the request (e.g. timeouts or errors from deeper layers etc..)